wandb/integration/ultralytics/callback.py

```diff
@@ -0,0 +1,367 @@
+            classification_columns = [
+                "Image",
+                "Predicted-Category",
+                "Prediction-Confidence",
+                "Top-5-Prediction-Categories",
+                "Top-5-Prediction-Confindence",
+                "Probabilities",
+                "Speed",
+            ]
+            validation_columns = ["Data-Index", "Batch-Index"] + classification_columns
+            validation_columns.insert(3, "Ground-Truth-Category")
+            train_columns = ["Epoch"] + validation_columns
+            self.train_validation_table = wandb.Table(columns=train_columns)
+            self.validation_table = wandb.Table(columns=validation_columns)
+            self.prediction_table = wandb.Table(columns=classification_columns)
+        elif model.task == "pose":
+            validation_columns = [
+                "Data-Index",
+                "Batch-Index",
+                "Image-Ground-Truth",
+                "Image-Prediction",
+                "Num-Instances",
+                "Mean-Confidence",
+                "Speed",
+            ]
+            train_columns = ["Epoch"] + validation_columns
+            self.train_validation_table = wandb.Table(columns=train_columns)
+            self.validation_table = wandb.Table(columns=validation_columns)
+            self.prediction_table = wandb.Table(
+                columns=[
+                    "Image-Prediction",
+                    "Num-Instances",
+                    "Mean-Confidence",
+                    "Speed",
+                ]
+            )
+
+    def _make_predictor(self, model: YOLO):
+        overrides = model.overrides.copy()
+        overrides["conf"] = 0.1
+        self.predictor = TASK_MAP[model.task][3](overrides=overrides, _callbacks=None)
+
+    def _save_model(self, trainer: DetectionTrainer):
+        model_checkpoint_artifact = wandb.Artifact(f"run_{wandb.run.id}_model", "model")
+        checkpoint_dict = {
+            "epoch": trainer.epoch,
+            "best_fitness": trainer.best_fitness,
+            "model": copy.deepcopy(de_parallel(self.model)).half(),
+            "ema": copy.deepcopy(trainer.ema.ema).half(),
+            "updates": trainer.ema.updates,
+            "optimizer": trainer.optimizer.state_dict(),
+            "train_args": vars(trainer.args),
+            "date": datetime.now().isoformat(),
+            "version": __version__,
+        }
+        checkpoint_path = trainer.wdir / f"epoch{trainer.epoch}.pt"
+        torch.save(checkpoint_dict, checkpoint_path, pickle_module=pickle)
+        model_checkpoint_artifact.add_file(checkpoint_path)
+        wandb.log_artifact(
+            model_checkpoint_artifact, aliases=[f"epoch_{trainer.epoch}"]
+        )
+
+    def on_train_start(self, trainer: DetectionTrainer):
+        with telemetry.context(run=wandb.run) as tel:
+            tel.feature.ultralytics_yolov8 = True
+
+    def on_fit_epoch_end(self, trainer: DetectionTrainer):
```

Comment:
Shouldn't the `trainer` be of type `Union[DetectionTrainer, ,PoseTrainer, ClassificationTrainer]` because the methods implements for these.